core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -307,7 +307,10 @@ public static Schema schemaFor(Table table, long snapshotId) {
       return schema;
     }
 
-    // TODO: recover the schema by reading previous metadata files
+    snapshot = table.snapshot(snapshot.parentId());
+    if (snapshot != null) {
+      return table.schemas().get(snapshot.schemaId());
```

Comment:
[question] can it be possible that even parentSnapshot is also lacking schemaId ? 

If this is true we can now return `null` earlier we were returning `table.schema()` instead. WDYT ?